### PR TITLE
Improve native_batch_norm_backward performance (CUDA)

### LINKED ISSUE #38915: Different speed between BatchNorm1d and BatchNorm2d
Hi I assume BatchNorm1d and BatchNorm2d (or 3d) should have same performance. However when normalizing a tensor of shape `(L, C) = (B*H*W, C)` by BatchNorm1d, the speed is >5x slower than BatchNorm2d on a shape of `(B, C, H, W)`. Is this expected or is there a way to speed up?

Reproducer:

```
import torch

def time_bn(bn, x):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    x = bn(x)
    end.record()
    torch.cuda.synchronize()
    print(start.elapsed_time(end))
    return x

shape = (50, 50, 50, 50)
a = torch.rand(shape)
b = a.flatten(0, -2)

bna = torch.nn.BatchNorm2d(50)
bnb = torch.nn.BatchNorm1d(50)

time_bn(bna, a)
time_bn(bnb, b)
```

Output:
```
11.409407615661621
89.55596923828125
```

## Environment

```
 - PyTorch Version: tested on 1.4/1.5
 - OS: Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.7/3.8
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: GTX 1070
 - Any other relevant information:
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel @csarofeen @ptrblck @xwang233 @albanD @mruberry @VitalyFedyunin

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes  #38915

The original code uses a single kernel to do both the reduction and the elementwise backward calculations. Whereas the  `SyncBatchNorm` kernels are split, which makes them slightly slower in some cases. I try to use the fused kernel when it's beneficial, but otherwise choose the optimized channels last split kernels. There is also eval mode, where the reduction is sometimes unnecessary in which case split kernels are a win even without channels last.

Benchmarks on my system show significant speedups for channels last reductions and eval mode, with only a few minor slowdowns in training mode. These slowdowns are for 2 x 2048 shape in training, which is a small channels last inputs. But for larger batches or channels, the channels last kernels are much faster.

|N   |C   |L   |training|backward|old   |new   |cudnn |
|----|----|----|--------|--------|------|------|------|
|1   |256 |3136|TRUE    |all     |70.25 |64.93 |68.90 |
|    |    |    |TRUE    |self    |69.77 |64.61 |69.42 |
|    |    |    |FALSE   |all     |70.10 |51.12 |x     |
|    |    |    |FALSE   |self    |70.17 |51.17 |x     |
|3136|256 |    |TRUE    |all     |554.08|76.63 |549.88|
|    |    |    |TRUE    |self    |553.34|78.19 |552.36|
|    |    |    |FALSE   |all     |565.40|55.09 |x     |
|    |    |    |FALSE   |self    |565.71|54.84 |x     |
|2   |8192|1   |TRUE    |all     |155.47|47.26 |202.26|
|    |    |    |TRUE    |self    |155.46|48.36 |203.72|
|    |    |    |FALSE   |all     |178.28|40.90 |x     |
|    |    |    |FALSE   |self    |178.21|40.69 |x     |
|2   |2048|1   |TRUE    |all     |43.50 |48.21 |57.47 |
|    |    |    |TRUE    |self    |43.63 |47.24 |55.22 |
|    |    |    |FALSE   |all     |49.36 |39.27 |x     |
|    |    |    |FALSE   |self    |49.25 |42.02 |x     |
|128 |8192|1   |TRUE    |all     |762.70|106.45|336.52|
|    |    |    |TRUE    |self    |765.79|107.04|337.32|
|    |    |    |FALSE   |all     |792.68|74.94 |x     |
|    |    |    |FALSE   |self    |793.86|74.83 |x     |
|128 |2048|1   |TRUE    |all     |188.37|46.20 |85.02 |
|    |    |    |TRUE    |self    |188.47|47.57 |85.04 |
|    |    |    |FALSE   |all     |191.57|40.44 |x     |
|    |    |    |FALSE   |self    |190.13|41.55 |x     |
|2   |8192|    |TRUE    |all     |156.03|43.01 |155.19|
|    |    |    |TRUE    |self    |156.24|46.59 |156.93|
|    |    |    |FALSE   |all     |179.34|40.06 |x     |
|    |    |    |FALSE   |self    |179.20|41.85 |x     |
|2   |2048|    |TRUE    |all     |44.05 |50.15 |44.21 |
|    |    |    |TRUE    |self    |44.10 |48.97 |44.11 |
|    |    |    |FALSE   |all     |49.72 |40.95 |x     |
|    |    |    |FALSE   |self    |49.87 |43.43 |x     |
|128 |8192|    |TRUE    |all     |775.19|96.60 |777.64|
|    |    |    |TRUE    |self    |776.20|96.85 |774.21|
|    |    |    |FALSE   |all     |797.64|68.01 |x     |
|    |    |    |FALSE   |self    |806.25|68.05 |x     |
|128 |2048|    |TRUE    |all     |188.49|48.10 |188.97|
|    |    |    |TRUE    |self    |188.07|46.97 |187.98|
|    |    |    |FALSE   |all     |192.32|43.78 |x     |
|    |    |    |FALSE   |self    |193.72|40.82 |x     |

### CODE REVIEW

**Comment by ngimel** [comment]
Awesome! Can you add benchmarks similar to original issue, very large N and small C, (125000, 5000) or something like that?

**Comment by peterbell10** [comment]
Sure, here are the shapes from the issue
|shape|old (us)|new (us)|
|-|-|-|
|(125000, 50)|6440|639|
|(50, 50, 50, 50)|1100|1090|

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 1]
```diff
@@ -1468,7 +1581,7 @@ at::Tensor batch_norm_backward_elemt_channels_last_cuda_template(
   if (input.scalar_type() == at::kHalf && weight.defined() && weight.scalar_type() == at::kFloat) {
     AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, input.scalar_type(), "batchnorm_backward_element", [&] {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1582]: pre-existing condition, but why do we need dispatch macro here (that would compile the kernel for all possible dtypes), when `input.scalar_type` is guaranteed to be at::kHalf?
Also, how come there's special accscalar_t handling for half, but not for bfloat16?

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 2]
```diff
@@ -1512,4 +1625,62 @@ at::Tensor batch_norm_backward_elemt_channels_last_cuda_template(
   return grad_input;
 }
 
+at::Tensor batch_norm_backward_elemt_channels_last_cuda_template(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1628]: this is, mostly, same function as above, but without `count` argument?

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 3]
```diff
@@ -1512,4 +1625,62 @@ at::Tensor batch_norm_backward_elemt_channels_last_cuda_template(
...
+    using accscalar_t = at::acc_type<scalar_t, true>;
+
+    if (weight.defined() && weight.scalar_type() != input.scalar_type()) {
+      batch_norm_backward_elemt_channels_last_kernel<ELEMENTS_PER_ITER>
+          <<<grid, block, 0, stream>>>(
+          grad_output.data_ptr<scalar_t>(),
+          input.data_ptr<scalar_t>(),
+          mean.data_ptr<accscalar_t>(),
+          inv_std.data_ptr<accscalar_t>(),
+          weight.defined() ? weight.data_ptr<accscalar_t>() : nullptr,
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1658]: nit: weight is guaranteed to be defined

**Review by ngimel — COMMENTED** [review]
Great work, @peterbell10 , I have a few questions about the handling of unfortunately large number of weight/bias type and input type combinations (our testing for those is probably not sufficient).

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 4]
```diff
@@ -106,6 +116,125 @@ void batch_norm_elementwise(
...
+        .add_input(invstd_nd)
+        .add_input(weight_nd)
+        .check_all_same_dtype(false)
+        .promote_inputs_to_common_dtype(false)
+        .build();
+
+    AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, grad_out.scalar_type(),
+                                    "batch_norm_eval_backward", [&]{
+      using accscalar_t = at::acc_type<scalar_t, true>;
+      gpu_kernel(iter, [] GPU_LAMBDA (scalar_t gO, accscalar_t invstd, accscalar_t weight)
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cu:225]: I believe that goes through the dynamic casting kernel.

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 5]
```diff
@@ -284,35 +413,72 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_cuda(const Tensor& self, const c10
...
+  const bool needs_reduction = train || grad_input_mask[1] || grad_input_mask[2];
+
+  // Fused reducion & elementwise kernel
+  if (needs_reduction && grad_input_mask[0] &&
+      !batch_norm_use_channels_last_kernels(input) &&
+      cuda::detail::canUse32BitIndexMath(input) &&
+      cuda::detail::canUse32BitIndexMath(grad_out)) {
+    return AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, input.scalar_type(),
+                                           "batch_norm_backward_cuda", [&] {
+      return batch_norm_backward_cuda_template<scalar_t, scalar_t, int32_t>(
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cu]: Fixed and added a test for mixed dtype batch_norm.

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 6]
```diff
@@ -391,33 +557,39 @@ std::tuple<Tensor, Tensor> batch_norm_gather_stats_with_counts_cuda(
...
   const Tensor& weight = *weight_maybe_owned;
 
-  // self is grad_output
-  if (at::cuda::detail::canUse32BitIndexMath(self) && batch_norm_use_channels_last_kernels(self)){
-    return batch_norm_backward_reduce_cuda_channels_last_template(self, input, mean, invstd, weight, input_g, weight_g, bias_g);
+  if (at::cuda::detail::canUse32BitIndexMath(grad_output) &&
+      batch_norm_use_channels_last_kernels(grad_output) &&
+      batch_norm_use_channels_last_kernels(input) &&
+      (!weight.defined() || weight.is_contiguous()) &&
+      mean.is_contiguous() && invstd.is_contiguous()){
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cu:590]: The kernel just calls `data_ptr` and assumes the layout without ever looking at the strides. So you could get incorrect results before.

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 7]
```diff
@@ -391,33 +557,39 @@ std::tuple<Tensor, Tensor> batch_norm_gather_stats_with_counts_cuda(
...
-    bool is_bfloat16_float = std::is_same<scalar_t, at::BFloat16>::value && mean_st == at::kFloat;
-    if (cuda::detail::canUse32BitIndexMath(self)) {
-      if (is_half_float || is_bfloat16_float) {
-        return batch_norm_backward_reduce_cuda_template<scalar_t, float, int32_t>(self, input, mean, invstd, weight, input_g, weight_g, bias_g);
+    const bool is_half_float = weight.defined() &&
+        weight.scalar_type() != input.scalar_type();
+    using accscalar_t = at::acc_type<scalar_t, true>;
+
+    if (cuda::detail::canUse32BitIndexMath(grad_output)) {
+      if (is_half_float) {
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cu]: Renamed the variable to `is_mixed_type` since this covers BFloat16 as well.

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 8]
```diff
@@ -791,16 +816,56 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> batch_norm_backward_reduce_cuda_templ
...
+  batch_norm_backward_elemt_kernel<input_scalar_t, stat_scalar_t, stat_accscalar_t, index_t>
+      <<<blocks_trans, threads_trans, 0, stream>>>
+      (input, grad_output, mean, invstd, weight, sum_dy, sum_dy_xmu, grad_input, norm_fct);
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+  return grad_input_reshaped.view(input_.sizes());
+}
+
+template<typename input_scalar_t, typename stat_scalar_t, typename index_t>
+Tensor batch_norm_backward_elemt_cuda_template(const Tensor& grad_out_, const Tensor& input_,
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cuh:860]: I suppose `count` could be an optional argument, but having them as two functions allows me to only compile the int32 kernel in the non-`count` case.

File: aten/src/ATen/native/cuda/Normalization.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/cuda/Normalization.cu:225]: Ugh, dynamic casting kernel is noticeably slower, but it probably doesn't happen often enough to warrant separate instantiation.

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 9]
```diff
@@ -822,6 +887,7 @@ Tensor batch_norm_backward_elemt_cuda_template(const Tensor& grad_out_, const Te
   int tb = std::max<int>(64/tf, 1);
   dim3 blocks_trans(input.size(1), std::max<int>(1, std::min<int>((256*1024)/input.size(1),
                                                                   (input.size(0)+tb-1)/tb)));
+  blocks_trans.y = std::min(blocks_trans.y, MAX_GRID_SIZE);
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cuh:890]: Yes, I think SyncBatchNorm with a large batch and only 1 channel would cause a cuda launch error. None of the tests hit this condition, but the same bug existed with the forward kernels (which use the same calculation) and one of the tests caught it there during the last PR.

File: aten/src/ATen/native/cuda/Normalization.cuh (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cuh]: Fixed BFloat16 handling and cleaned this up a bit.

File: aten/src/ATen/native/cuda/Normalization.cuh (hunk 2 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cuh:1630]: Yes, `count` is only used by `SyncBatchNorm`.

**Comment by peterbell10** [comment]
@ngimel PTAL

**Review by ngimel — APPROVED** [review]
Great, thanks for the quick turnaround
